Symmetrize fourth order tensor

vinayakvijay opened the pull request:
Closes #17403. FYI @bangerth. I have extended SymmetricTensor::symmetrize() for fourth-order tensors and have added a test to test the same.

vinayakvijay commented:
I have changed the status to draft.  @bangerth there are two things i would like to have some comments on :

1. What should be the type of the variable `symmetry_type`? Right now I have considered it to be of type `unsigned int`. For only minor symmetry its value should be zero and, for major and minor symmetry its value should be 1. Now, if we want only these symmetry types then I would maybe like the type to be changed to `bool`. But if we want other permutations of the symmetry types (which I feel complicates things a bit much) the we would maybe want to keep it `unsigned int` or `string`. This is, basically, my next doubt.
2. Do we want symmetry types of the type $$A_{ijkl}=A_{jikl}$$ - that is mapping tensors to symmetric tensors. The vice versa case is the other type. And then, there is the question of having the major symmetry, as well, with these symmetries. This adds to the possible cases for the argument `symmetry_type`.

masterleinad commented:
Any good reason not to use an enum (class)?

vinayakvijay commented:
> Any good reason not to use an enum (class)?

I hadn't considered it. Yes, I think it is a good option.

bangerth commented on include/deal.II/base/symmetric_tensor.h:
```diff
@@ -3544,6 +3544,121 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+ * An enumeration to describe the desired symmetry of the
+ * symmetric fourth order tensor
+ */
+enum class SymmetryType
```
Would you be willing to rename this to `TensorSymmetryType` to make it more specific to just tensors? Symmetry is a concept that appears in many places, and I'm concerned about using an overly generic name.

bangerth commented on include/deal.II/base/symmetric_tensor.h:
```diff
@@ -3544,6 +3544,121 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
 }
 
+/**
+ * An enumeration to describe the desired symmetry of the
+ * symmetric fourth order tensor
```
"the symmetric fourth order tensor" assumes that the reader is already familiar with what operation you are talking about. But there is no reference to any operation here (or to a function that uses this enum). Would you mind adding this piece of information? Perhaps in the following way:
```suggestion
 * An enumeration to describe the desired symmetry one wants to obtain in
 * symmetrizing a fourth order tensor. This enum class is used by the symmetrize()
 * function.
```

bangerth commented on include/deal.II/base/symmetric_tensor.h:
```diff
@@ -3544,6 +3544,121 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+/**
+ * Return the symmetrized version of a full rank-4 tensor, i.e.
+ * as a symmetric rank-4 tensor.
+ * @param symmetry_type decides the type of symmetry.
```
Doxygen shows the name of the argument separate from the text that follows. So perhaps this?
```suggestion
 * @param symmetry_type This argument decides the type of symmetry.
```

bangerth commented on include/deal.II/base/symmetric_tensor.h:
```diff
@@ -3544,6 +3544,121 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+ * Return the symmetrized version of a full rank-4 tensor, i.e.
+ * as a symmetric rank-4 tensor.
+ * @param symmetry_type decides the type of symmetry.
+ * This is the version for general dimensions.
```
I think this line should be part of the general text at the top, but as is it will be shown as part of the `@param` paragraph.

bangerth commented on include/deal.II/base/symmetric_tensor.h:
```diff
@@ -3544,6 +3544,121 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+
+  const Number half = internal::NumberType<Number>::value(0.5);
+
+  if (symmetry_type == SymmetryType::minor)
```
This would lend itself to a `switch`-`case` statement. In any case, you should add a `default` case (or, in the current `if-else if` structure, a final `else`) that simply says `DEAL_II_NOT_IMPLEMENTED()` to capture the case where someone passes in an unsupported case for the second argument.

bangerth commented on include/deal.II/base/symmetric_tensor.h:
```diff
@@ -3544,6 +3544,121 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+  SymmetricTensor<4, dim, Number>
+  symmetrize(const Tensor<4, dim, Number> &t,
+             const SymmetryType &symmetry_type = SymmetryType::minor)
```
Have you considered not giving this argument a default value? It's not clear to me that one or the other choice for this argument is "more natural" and so should be the default. It might be better to just *always* require the caller to say what it is that they want, if only to force them to think about what that is (and thereby to also ensure that their choice is documented).

vinayakvijay replied:
Thanks for your comment. Now that I think about it, the current  design of SymmetricTensor class _requires_ minor symmetry by default. So the only additional symmetry we can have is the major symmetry. This means that we don't even need the proposed enum class TensorSymmetryType. We could simply have a bool argument for additional major symmetry instead of the TensorSymmetryType argument.

bangerth replied:
Yes, I missed that too. This sounds like the right approach then!

bangerth commented on include/deal.II/base/symmetric_tensor.h:
```diff
@@ -3544,6 +3544,121 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+  const Number half = internal::NumberType<Number>::value(0.5);
+
+  if (symmetry_type == SymmetryType::minor)
+    {
+      // minor symmetry - A_{ijkl}=A_{jikl}=A_{ijlk}=A_{jilk}
+
+      for (unsigned int i = 0; i < dim; ++i)
+        for (unsigned int j = 0; j < dim; ++j)
+          for (unsigned int k = 0; k < dim; ++k)
+            for (unsigned int l = 0; l < dim; ++l)
+              {
+                if (i != j && k == l)
+                  {
+                    // A_{ijkk}=A_{jikk}
+                    result[i][j][k][k] = (t[i][j][k][k] + t[j][i][k][k]) * half;
+                  }
+                else if (i == j && k != l)
+                  {
+                    // A_{iikl}=A_{iilk}
+                    result[i][i][k][l] = (t[i][i][k][l] + t[i][i][l][k]) * half;
+                  }
+                else if (i != j && k != l)
+                  {
+                    // A_{ijkl}=A_{jilk}
+                    result[i][j][k][l] = (t[i][j][k][l] + t[j][i][k][l] +
+                                          t[i][j][l][k] + t[j][i][l][k]) *
+                                         half * half;
+                  }
+                else
+                  {
+                    // A_{iijj} unchanged
+                    result[i][j][k][l] = t[i][j][k][l];
+                  }
+              }
```
Alternatively, if I see this right, you duplicate this code below, right? If that's so, you can just *always* execute this part, and only do something *in addition* for the major symmetry.

bangerth reviewed (commented):
@vinayakvijay I'm sorry for letting this fall through the crack. I believe that your last update happened just before the deal.II users workshop, and then it just got cycled down the page of open pull requests :-(

Below are some comments that should not be very difficult to address.

vinayakvijay commented:
@bangerth I have addressed the comments. Request you to have a look.

bangerth commented on include/deal.II/base/symmetric_tensor.h:
```diff
@@ -3546,6 +3546,67 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+/**
+ * Return the symmetrized version of a full rank-4 tensor, i.e.
+ * as a symmetric rank-4 tensor. This is the version for general dimensions.
+ * @param major_symmetry  This argument decides the presence of major symmetry.
```
Since the `enum` has now disappeared, you no longer document what "major" and "minor" symmetry means. Would you mind re-adding that here?

bangerth commented on include/deal.II/base/symmetric_tensor.h:
```diff
@@ -3546,6 +3546,67 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+template <int dim, typename Number>
+DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+  SymmetricTensor<4, dim, Number>
+  symmetrize(const Tensor<4, dim, Number> &t, const bool &major_symmetry)
```
```suggestion
  symmetrize(const Tensor<4, dim, Number> &t, const bool major_symmetry)
```

bangerth reviewed (commented):
Yes, almost there!

vinayakvijay commented:
I have incorporated the comments. There's one more thing: I wasn't able to indent using `./contrib/utilities/indent` because of the first name last name issue. This could be because I had only the first name for my initial commits. I have corrected this for my latest commit. Still, not able to indent using the script though.

bangerth commented:
I think the easiest way to deal with this is if you were to squash or fix-up your commits into one. Then you can do
```
  git --amend --author="...your name and email..."
```
to get these fields fixed as well.

vinayakvijay commented:
Thanks. I have corrected the authorship. Hope it is fine now.

bangerth commented:
/rebuild

bangerth commented:
There are still two failing checks:
* "/home/runner/work/dealii/dealii/include/deal.II/base/symmetric_tensor.h:3465: warning: The following parameter of symmetrize(const Tensor< 4, dim, Number > &t, const bool major_symmetry) is not documented: parameter 't'". What this says is that doxygen wants you to document *all* parameters with a `@param` tag if you document *any* of them. Since you're already documenting the second parameter, just also add a `@param` tag for the first, saying something like "@param t The tensor to be symmetrized."
* The indent check wants you to break an overly long line, see https://github.com/dealii/dealii/actions/runs/10807262527/job/30001651784?pr=17408

bangerth commented:
The indent script is still unhappy... https://github.com/dealii/dealii/actions/runs/10829382016/job/30113752867?pr=17408

bangerth commented:
OK, assuming this works now, may I suggest one last change? There are currently 7 commits in this pull request, but the majority really just fix up things that went wrong in the previous ones. May I suggest you "squash" (or better even, "fix up") these commits into one? This has the advantage that anyone later going through the history will only encounter one commit that actually lets all tests (including the indentation and doxygen tests) succeed. Having all commits have all tests work is a useful invariant when trying to figure later on which test broke something.

bangerth commented:
Yay!
